File: N-Queens.java
/*
 * LeetCode 51 N-Queens
 */
class Solution {
    public List<List<String>> solveNQueens(int n) {
        char[][] board = new char[n][n];
        for (char[] chars : board) {
            Arrays.fill(chars,'.');
        }
        List<List<String>> res = new ArrayList<>();
        backtrack(res,0,board);
        return res;
    }

    public void backtrack(List<List<String>> res,int row,char[][] board) {
        if (row == board.length) {
            res.add(buildList(board));
            return ;
        }
        for (int i = 0; i < board.length; i++) {
            if (!isValid(board,row,i)) {
                continue;
            }
            board[row][i] = 'Q';
            backtrack(res,row + 1,board);
            board[row][i] = '.';
        }
    }

    public boolean isValid(char[][] board,int r, int c) {
        int n = board.length;
        for (int i = 0; i < n; i++) {
            if (board[i][c] == 'Q') {
                return false;
            }
        }
        for (int i = r - 1,j = c + 1; i >= 0 && j < n;i--,j++) {
            if (board[i][j] == 'Q') {
                return false;
            }
        }
        for (int i = r - 1,j = c - 1; i >= 0 && j >= 0;i--,j--) {
            if (board[i][j] == 'Q') {
                return false;
            }
        }
        return true;
    }

    private List<String> buildList(char[][] chess) {
        List<String> path = new ArrayList<>();
        for (int i = 0; i < chess.length; i++) {
            path.add(new String(chess[i]));
        }
        return path;
    }
}

File: N-Queens.go

var res [][]string

func solveNQueens(n int) [][]string {
	res = [][]string{}
	queens := make([]int,n)
	for i := 0;i < n;i++ {
		queens[i] = -1
	}
	columns,ldiag,rdiag := map[int]bool{},map[int]bool{},map[int]bool{}
	backtrack(queens,n,0,columns,ldiag,rdiag)
	return res
}

func backtrack(queens []int,n int,row int,columns,ldiag,rdiag map[int]bool) {
	if row == n {
		pic := generateBoard(queens,n)
		res = append(res,pic)
		return
	}

	for i := 0;i < n;i++ {
		ldiagIdx := row - i
		rdiagIdx := row + i
		if columns[i] || ldiag[ldiagIdx] || rdiag[rdiagIdx] {
			continue
		}

		queens[row] = i
		columns[i],ldiag[ldiagIdx],rdiag[rdiagIdx] = true,true,true
		backtrack(queens,n,row + 1,columns,ldiag,rdiag)
		queens[row] = -1
		delete(columns,i)
		delete(ldiag,ldiagIdx)
		delete(rdiag,rdiagIdx)


	}
}


func generateBoard(queens []int,n int) []string {
	pic := []string{}
	for i := 0;i < n;i++ {
		row := make([]byte,n)
		for j := 0;j < n;j++ {
			row[j] = '.'
		}
		row[queens[i]] = 'Q'
		pic = append(pic,string(row))
	}
	return pic
}
